File: C_EXAMPLE/example_basic_Brake.cpp
/**
 * @file example_basic_Brake.cpp
 * @brief Example demonstrating how to control the brakes of L1 robot arms.
 *
 * This example shows how to:
 * - Query SDK and controller versions
 * - Check and recover the robot arm state (IDLE / ERROR)
 * - Unlock and lock all brakes of a specified robot arm
 *
 * @warning Releasing the brakes may cause the arm to drop due to gravity.
 *          Ensure the arm is held securely before unlocking the brakes.
 */

#include "L1Robot.h"

/**
 * @brief Entry point of the brake control example.
 *
 * Workflow overview:
 * 1. Initialize communication with the robot controller
 * 2. Retrieve SDK and controller versions
 * 3. Verify the arm state and recover to IDLE if necessary
 * 4. Unlock all brakes of ARM0
 * 5. Lock all brakes of ARM0
 *
 * @param[in] argc Argument count (unused)
 * @param[in] argv Argument vector (unused)
 * @return int Exit code (0 on normal exit)
 */
int main(int argc, char** argv)
{
    int sdk_version = 0;                  ///< SDK version number
    int controller_version = 0;           ///< Controller firmware version
    unsigned int system_errorcode = 0;    ///< Last system error code
    FXStateType obj_state = FX_STATE_UNKNOWN;

    /* Get SDK version */
    sdk_version = FX_L1_System_GetSDKVersion();
    printf("SDK version is 0x%08x\n", sdk_version);

    /* Establish communication with the controller */
    if (FX_L1_System_Link(6, 6, 7, 190, FX_LOG_ALL_FLAG) < 0)
    {
        printf("Failed to link system\n");
        goto WAIT_EXIT;
    }

    /* Get controller version */
    controller_version = FX_L1_System_GetControllerVersion();
    printf("Controller version is 0x%08x\n", controller_version);

    /*
     * Check current arm state.
     * Brake operations are only allowed in FX_STATE_IDLE.
     */
    obj_state = FX_L1_Fbk_CurrentState(FX_OBJ_ARM0);
    if (obj_state == FX_STATE_ERROR)
    {
        printf("Brake lock/unlock must be run in FX_STATE_IDLE state, try to reset error\n");
        if (FX_L1_State_ResetError(FX_OBJ_ARM0, 1000, &system_errorcode) == FUNC_RET_SUCCESS)
        {
            printf("Reset arm0 error success, arm0 is now in STATE_IDLE state\n");
        }
        else
        {
            printf("Failed to reset arm0 error, errorcode = 0x%08x\n", system_errorcode);
            goto WAIT_EXIT;
        }
    }
    else if (obj_state != FX_STATE_IDLE)
    {
        if (FX_L1_State_SwitchToIdle(FX_OBJ_ARM0, 1000) != FUNC_RET_SUCCESS)
        {
            printf("Failed to transfer arm0 to STATE_IDLE state\n");
            goto WAIT_EXIT;
        }
    }

    /* Unlock all brakes of ARM0 */
    printf("Press any key to unlock all brakes of arm0, "
           "please hold the arm carefully or it will drop!\n");
    getchar();
    if (FX_L1_Config_SetBrakeUnlock(FX_OBJ_ARM0, 0xFF) != FUNC_RET_SUCCESS)
    {
        printf("Unlock all brakes of arm0 failed\n");
    }

    /* Lock all brakes of ARM0 */
    printf("Press any key to lock all brakes of arm0\n");
    getchar();
    if (FX_L1_Config_SetBrakeLock(FX_OBJ_ARM0, 0xFF) != FUNC_RET_SUCCESS)
    {
        printf("Lock all brakes of arm0 failed\n");
    }

WAIT_EXIT:
    printf("Press any key to exit\n");
    getchar();
    return 0;
}

File: PYTHON_EXAMPLE/example_basic_Brake.py
"""
@file example_basic_Brake.py
@brief Example demonstrating how to control the brakes of L1 robot arms.

Workflow overview:
    1. Initialize communication with the robot controller
    2. Retrieve SDK and controller versions
    3. Verify the arm state and recover to IDLE if necessary
    4. Unlock all brakes of ARM0
    5. Lock all brakes of ARM0

@warning Releasing the brakes may cause the arm to drop due to gravity.
         Ensure the arm is held securely before unlocking the brakes.

@warning Press Enter key at any time to stop the robot and exit.
"""

import sys
import time
import threading
from pathlib import Path

root_dir = Path(__file__).parent.parent
if str(root_dir) not in sys.path:
    sys.path.insert(0, str(root_dir))
from PYTHON_SDK.GentoRobot import GentoRobot, FXLogMask, FXObjMask, FXObjType, state_map,error_dict

def main():

    ctrl_obj=FXObjType.OBJ_ARM0
    axis_mask = 0x7F
    arm_idx=0
    if ctrl_obj==FXObjType.OBJ_ARM1:
        arm_idx=1
    log_mask=FXLogMask.FX_LOG_INFO_FLAG

    robot = GentoRobot()

    print(f"\n### 1/6 .Connecting...")
    ret = robot.link(6, 6, 7, 190, log_level=log_mask)
    if not robot._connected:
        print(f"Link failed. Error msg: {robot._get_operate_error_msg(ret)}")
        return

    print(f"\n### 2/6. Getting versions...")
    print(f"Controller version:{robot.get_controller_version()}")
    print(f"Sdk version:{robot.get_sdk_version()}")

    print(f"\n### 3/6. IDLE state switching...")
    arm0_state=robot.current_state(ctrl_obj)
    if state_map[arm0_state] == "Error":
        ret, system_errorcode = robot.reset_error(ctrl_obj, 2000)
        if ret != 0:
            print(f"Reset error failed. Error msg: {error_dict[system_errorcode]}")
            return
    if state_map[arm0_state] != "IDLE":
        ret = robot.switch_to_idle(ctrl_obj, 2000)
        if ret != 0:
            print(f"Switch to IDLE failed. Error msg: {robot._get_operate_error_msg(ret)}")
            return
        rt_dict = robot.get_rt_dict()
        print(f"current state：{rt_dict["arms"][arm_idx]["state"]["cur"]}")

    print(f"\n### 4/6. Unlock all brakes...")
    input("Press Enter key to unlock all brakes of arm0, please hold the arm carefully or it will drop!")
    ret=robot.config_brake_unlock(ctrl_obj, axis_mask)
    if ret != 0:
        print(f"Unbrake failed, Error msg: {robot._get_operate_error_msg(ret)}")
        return

    print(f"\n### 5/6. Lock all brakes...")
    input("Press Enter key to lock all brakes of arm0")
    ret = robot.config_brake_lock(ctrl_obj, axis_mask)
    if ret != 0:
        print(f"Brake failed, Error msg: {robot._get_operate_error_msg(ret)}")
        return

    print(f"\n### 6/6. Task finished.")

if __name__ == "__main__":
    main()
